### PR TITLE
JSON: support pretty JSON formatting

### PR DESCRIPTION
**Pull request by timbertson**
I've made an attempt at pretty JSON rendering (reusing some of the YAML internals), i.e. fixing #932.

~This is not the final version, I've basically just put it into renderYAML in order to minimise the diff for review. If it's 👍 , I'm expecting we'd move it into a new (say) `renderTo` function which takes the Format argument, and `renderYAML` / `renderJSON` would become trivial re-exports of that.~

You can see the new formatting in the inline tests.

### Compact rendering

I haven't retained support for compact JSON rendering, but could do so if that's deemed a useful feature.

### ~Comma placement~ (fixed now)

The main thing I'm not pleased with is the comma placement, for ease of implementation I ended up with unaligned leading commas, e.g.:

```
{
  "zero": []
  ,"one": [ "a" ]
  ,"two": [
( ... )
```

I did make a version which adds trailing commas to all-but-the-last-field, but the implementation was pretty awkward: https://github.com/timbertson/dhall-lang/commit/a8fd52f803990e8942bbd21cd7669271d92af7ff~

### CODE REVIEW

**Comment by timbertson** [comment]
I also thought about using the opening brace to align with commas, i.e.:

```
{ foo:
  [ 1
  , 2
  , 3
}
```

But couldn't quite crack the logic of when to add a newline. Maybe it's easy though, and my brain just wasn't working ;)

**Comment by Gabriel439** [comment]
@timbertson: I think formatting in the same style as `jq` is what will probably make the most people happy

**Comment by timbertson** [comment]
That matches what the linked https://github.com/timbertson/dhall-lang/commit/a8fd52f803990e8942bbd21cd7669271d92af7ff does, at the cost of some ugly implementations for `NonEmpty/{mapLeading,mapLast}`

**Comment by Gabriel439** [comment]
@timbertson: I think the additional complexity for the latter version would be fine

**Comment by sjakobi** [comment]
I agree that the new JSON rendering is very pretty! :+1:

**Comment by timbertson** [comment]
Thanks for the feedback, I got a slightly nicer version working by implementing `splitAt` instead of `splitLast`. Which could go into `List.splitAt` if you like it.

I've now pulled it out into a new `renderAs` function and `Format` type, (neither of which are exposed in JSON/package.dhall though). So I think it's ready to merge now.

Oh, it also includes #934 incidentally, so I guess that might cause a minor merge conflict (but if we merge that first, we can just take this version since it acts the same).

**Comment by timbertson** [comment]
(I'm a little confused by what the hydra failures are trying to tell me, and how to fix it.)

**Comment by sjakobi** [comment]
> (I'm a little confused by what the hydra failures are trying to tell me, and how to fix it.)

It says that the inferred type of the Prelude `package.dhall` doesn't exactly match `tests/type-inference/success/preludeB.dhall`. You can update that file by running `dhall type` on `tests/type-inference/success/preludeA.dhall`.

**Comment by Gabriel439** [comment]
@timbertson: If you have Nix installed, `./scripts/generate-test-files.sh` will take care of generating that (and other files) correctly

**Comment by timbertson** [comment]
I use nix everyday, but that script is trying to build 392 packages (currently building glibc :cold_sweat: ). I do have the hydra channel installed as a channel named `dhall`, can I can use that to get some more binary cache hits somehow?

**Comment by timbertson** [comment]
To the failures themselves, the current error is:

```
diff --recursive ./tests.actual/type-inference/success/preludeB.dhall ./tests.expected/type-inference/success/preludeB.dhall
229c229
<           ∀ ( j
---
>           ∀ ( json
builder for '/nix/store/vd4nvxwd8ig4jkg37zmr4khhr02pkz4f-test-files-lint.drv' failed with exit code 1
```

Which is because the argument is now named `json` instead of `j` (for `JSON.render`), I didn't actually change any _types_. I'm hoping it's OK to just manually edit that file for such a simple change?

(I tried @sjakobi's suggestion but it resulted in a huge diff, possibly because I'm running the latest dhall release instead of master)

**Comment by sjakobi** [comment]
> ...the argument is now named `json` instead of `j` (for `JSON.render`), I didn't actually change any _types_. I'm hoping it's OK to just manually edit that file for such a simple change?

Sure, a manual edit is fine if that fixes CI. Alternatively, you could just keep the old variable name.

If none of that works, I'd be interested in seeing that huge diff! :)

**Comment by Gabriel439** [comment]
@timbertson: Here are the instructions for using our cache: https://github.com/dhall-lang/dhall-haskell#nix

**Comment by timbertson** [comment]
The huge diff seems to just be due to how earlier paramaters are referred to, e.g.:

```
 , Function :
-    { compose :
-          ∀(A : Type)
-        → ∀(B : Type)
-        → ∀(C : Type)
-        → ∀(f : A → B)
-        → ∀(g : B → C)
-        → ∀(x : A)
-        → C
-    , identity : ∀(a : Type) → ∀(x : a) → a
+    { compose : Type → Type → Type → (_@2 → _@2) → (_@2 → _@2) → _@4 → _@3
+    , identity : Type → _ → _@1
     }
```

I'm guessing this has changed in `master` since last release

**Comment by Gabriel439** [comment]
@timbertson: Prefix the command with `XDG_CACHE_HOME=/var/empty` and the diff should get smaller

**Comment by sjakobi** [comment]
@timbertson I think if you use the `--no-cache` flag for `dhall type`, those variables will have the expected names. I can't quite remember how this worked though… :/ It's a bit of a gotcha…

**Comment by timbertson** [comment]
Thanks @Gabriel439 , that works (and prodices no diff from my cheeky manual edit :)).

So I believe this branch is now good to merge if you're happy with it, let me know if there's anything else to address.

File: Prelude/JSON/render [hunk 1]
```diff
@@ -7,51 +7,16 @@ let JSON =
...
-            ( JSON.array
-                [ JSON.bool True
-                , JSON.string "Hello"
-                , JSON.object
-                    [ { mapKey = "foo", mapValue = JSON.null }
-                    , { mapKey = "bar", mapValue = JSON.double 1.0 }
-                    ]
-                ]
-            )
-        ≡ "[ true, \"Hello\", { \"foo\": null, \"bar\": 1.0 } ]"
```
Comment (sjakobi) [inline Prelude/JSON/render]: Could you bring back this or another example? Examples also serve as documentation.

File: Prelude/JSON/renderYAML [hunk 2]
```diff
@@ -12,238 +12,16 @@ let JSON =
...
-let blockToText
-    : Block → Text
-    =   λ(block : Block)
-      → Text/concatMap
-          Text
-          (λ(line : Text) → line ++ "\n")
-          (NonEmpty/toList Text (Block/toLines block))
-
-let renderYAML
+let render
```
Comment (sjakobi) [inline Prelude/JSON/renderYAML]: Conventionally we've used the public names here.
```suggestion
let renderYAML
```

File: Prelude/JSON/renderYAML [hunk 3]
```diff
@@ -12,238 +12,16 @@ let JSON =
...
-        ≡ ''
-          ! "zero": []
-          ! "one":
-            - "a"
-          ! "two":
-            - "a"
-            - "b"
-          ''
-
-let example2 =
```
Comment (sjakobi) [inline Prelude/JSON/renderYAML]: One or multiple examples would be nice to have here.

File: tests/type-inference/success/preludeB.dhall [hunk 4]
```diff
@@ -226,7 +226,7 @@
             )
         → JSON
     , render :
-          ∀ ( j
+          ∀ ( json
```
Comment (sjakobi) [inline tests/type-inference/success/preludeB.dhall]: Where does this variable name change actually come from?

File: Prelude/JSON/Format [hunk 5]
```diff
@@ -0,0 +1 @@
+< YAML | JSON >
```
Comment (sjakobi) [inline Prelude/JSON/Format]: Could you add a bit of documentation?

File: Prelude/JSON/render (hunk 1 above)
Comment (timbertson) [reply Prelude/JSON/render]: They're still all there, in `renderAs`. That way each example gets rendered in both JSON and YAML, which it nice for consistency (but bad for locality). I've added a comment to both render & renderYAML for anyone seeking examples.

File: Prelude/JSON/renderYAML (hunk 2 above)
Comment (timbertson) [reply Prelude/JSON/renderYAML]: Good catch, this was accidental

File: tests/type-inference/success/preludeB.dhall (hunk 4 above)
Comment (timbertson) [reply tests/type-inference/success/preludeB.dhall]: `renderYAML` always took `json`, but `render` took `j`. Now they're both delegating to `renderAs`, which takes `json` (and that's what gets shown here)

File: Prelude/JSON/Format (hunk 5 above)
Comment (timbertson) [reply Prelude/JSON/Format]: Done :+1:

**Comment by timbertson** [comment]
> The sheer size of renderAs is a bit jarring

Yeah, I don't love that :/

FWIW it's only ~190 lines of plain code (excluding NonEmpty and examples). The previous code was 106 for yaml so it's at least less than double to also support JSON, though not by much. The amount of shared logic was disappointingly low between the two formats (they're subtly different in annoying ways), but there's still some benefit from the shared utilities (NonEmpty as well as render-specific helpers)

We could instead render YAML as pretty JSON and it would only be a net addition of about a dozen lines, but people didn't much like that thought in #932 :man_shrugging:

**Comment by timbertson** [comment]
> Admittedly I'm not a big fan of hiding the examples in that huge renderAs file, but maybe there are reasons to do that that I don't see.

I'm not too bothered, if you think it's worth splitting them up into their respective files? The main benefit I saw was that each new test case tests both outputs (if you're adding a test you should probably test both formats). Also it was easier to evaluate just one file for development, but that doesn't matter much for most people.

**Comment by sjakobi** [comment]
Oh, I'm not against having these tests in `renderAs` They make a lot of sense in that place. But for the purpose of documentation I think we could simply duplicate those examples to `render` and `renderYAML`. Sorry for not being clearer earlier!

**Comment by timbertson** [comment]
Oh I see, sure - easy enough :+1:. I've just duplicated one example, which I figure is enough for illustrative purposes.

**Comment by timbertson** [comment]
I've resolved merge conflicts against `master`

**Comment by Gabriel439** [comment]
@timbertson: Thank you! 🙂

**Pull request merged**